src/groth16/mod.rs

```diff
@@ -466,6 +771,170 @@ impl<'a, E: Engine> ParameterSource<E> for &'a Parameters<E> {
     }
 }
 
+// A re-usable method for parameter loading via mmap.
+fn read_g1<E: Engine>(
+    mmap: &Mmap,
+    start: usize,
+    end: usize,
+    checked: bool,
+) -> Result<E::G1Affine, std::io::Error> {
+    let ptr = &mmap[start..end];
+    let repr =
+        unsafe { *(ptr as *const [u8] as *const <E::G1Affine as CurveAffine>::Uncompressed) };
+
+    if checked {
+        repr.into_affine()
+    } else {
+        repr.into_affine_unchecked()
+    }
+    .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))
+    .and_then(|e| {
+        if e.is_zero() {
+            Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                "point at infinity",
+            ))
+        } else {
+            Ok(e)
+        }
+    })
+}
+
+// A re-usable method for parameter loading via mmap.
+fn read_g2<E: Engine>(
+    mmap: &Mmap,
+    start: usize,
+    end: usize,
+    checked: bool,
+) -> Result<E::G2Affine, std::io::Error> {
+    let ptr = &mmap[start..end];
+    let repr =
+        unsafe { *(ptr as *const [u8] as *const <E::G2Affine as CurveAffine>::Uncompressed) };
+
+    if checked {
+        repr.into_affine()
+    } else {
+        repr.into_affine_unchecked()
```

Comment:
I think this is actually not as useful as I hoped it would be. `into_affine_unchecked` and `into_affine` actually allocate new copies in memory, which means we are not actively using the mmaped data, but rather just reuse it to load the params.

Reply:
Ok. This code does allocate memory (as noted) because the params are not laid out on disk in such a way that we can just 'cast to struct' and use it.  The description of the PR is instead to allow lazy parameter loading, with the significant win (in terms of RAM usage) being that not all params are required to be in RAM all the time.  If that is not a useful goal, I agree, this does not help that much.